src/cli/upgrade.rs

```diff
@@ -344,3 +346,82 @@ fn ensure_package_exists(package_name: &str, available_packages: &[String]) -> m
     }
     .into())
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::Workspace;
+    use insta::assert_snapshot;
+    use std::io::Write;
+    use std::path::Path;
+    use tempfile::tempdir;
+
+    use super::*;
+
+    // When the specific template is not in the file or the file does not exist.
+    // Make the file and append the template to the file.
+    fn create_or_append_file(path: &Path, template: &str) -> std::io::Result<()> {
+        let file = fs_err::read_to_string(path).unwrap_or_default();
+
+        if !file.contains(template) {
+            std::fs::OpenOptions::new()
+                .append(true)
+                .create(true)
+                .open(path)?
+                .write_all(template.as_bytes())?;
+        }
+        Ok(())
+    }
+
+    // This test requires network connection, takes a lot of time to
+    // complete, and torch version can change over time, so we ignore
+    // it by default.
+    #[ignore]
```

Comment:
1. Do you think we could move this thing into integration tests (if possible), mark it as extra-slow. 
2. It might be benificial to take a slimmer dependency than torch:
`https://pypi.tuna.tsinghua.edu.cn/simple/` is an example public registry, if you take something like `click` I guess the same issue should occur as the linked one.
3. This option might also be relevant, to only take versions into account at time of test publish: https://pixi.sh/dev/reference/pixi_manifest/#exclude-newer-optional

wdyt?

Reply:
Thanks for suggestions! I applied all of them. Also, I tweaked types from `&Option<T>` to `Option<&T>`.

Reply:
I was thinking about introducing locally running registry server specifically for this test, but it seems not worth the hassle, so only changed package and index.

Reply:
I've been in that position at least 15 times in the last year :), I know the feeling. Would be great to do so at some point, but I am in agreement with what you changed.